src/main/scala/edu/berkeley/cs/rise/opaque/Utils.scala

```diff
@@ -293,18 +279,94 @@ object Utils extends Logging {
 
   var eid = 0L
   var attested: Boolean = false
-  var attesting_getepid: Boolean = false
-  var attesting_getmsg1: Boolean = false
-  var attesting_getmsg3: Boolean = false
-  var attesting_final_ra: Boolean = false
+  // Initialize accumulator variables for tracking the state of attestation
+  var acc_registered: Boolean = false
+  val numUnattested: LongAccumulator = new LongAccumulator
+  val numAttested: LongAccumulator = new LongAccumulator
 
   def initSQLContext(sqlContext: SQLContext): Unit = {
     sqlContext.experimental.extraOptimizations =
       (Seq(EncryptLocalRelation, ConvertToOpaqueOperators) ++
         sqlContext.experimental.extraOptimizations)
     sqlContext.experimental.extraStrategies = (Seq(OpaqueOperators) ++
       sqlContext.experimental.extraStrategies)
-    RA.initRA(sqlContext.sparkContext)
+
+    val sc = sqlContext.sparkContext
+    if (!acc_registered) {
+      sc.register(numUnattested)
+      sc.register(numAttested)
+      acc_registered = true
+    }
+
+    RA.wait(sc)
+    RA.run(sc)
+
+    val thread = new Thread {
+      override def run: Unit = {
+        while (true) {
+          RA.run(sc)
+        }
+      }
+    }
+    thread.start
+  }
+
+  def startEnclave(numUnattestedAcc: LongAccumulator) : Long = {
+    this.synchronized {
+      if (eid == 0L) {
+        val enclave = new SGXEnclave()
+        val path = findLibraryAsResource("enclave_trusted_signed")
+        eid = enclave.StartEnclave(path)
+        numUnattestedAcc.add(1)
+        logInfo("Starting an enclave")
+      }
+    }
+    eid
+  }
+
+  def initEnclave(): (SGXEnclave, Long) = {
+    // This following exception relies on the Spark fault tolerance and its retry mechanism
+    // in case of a task failure. Therefore, Spark MUST be configured to retry in case of a task failure
+    this.synchronized {
+      if (!attested) {
+        Thread.sleep(200)
+        throw new OpaqueException("Attestation not yet complete")
+      } else {
+        val enclave = new SGXEnclave()
+        (enclave, eid)
+      }
+    }
+
+  }  
+
+  def generateReport(): (Long, Option[Array[Byte]]) = {
+    this.synchronized {
+      if (!attested) {
```

Comment:
The solution I came up with involved adding:
```.scala
if (eid == 0) {
 Thread.sleep(200)
  throw new OpaqueException("Enclave has not been started")
}
```
and just letting the fault tolerance handle it. With this, there should be no need to have the `wait` function anymore.

Reply:
I changed this to not fail, and instead generate a report only when the enclave has been started.